### PR TITLE
Add OptimizeEpilogue pass.

### PR DESCRIPTION
**Pull request by oplavsic**
This commit is cherry-picked from upstream and updated to support MFMA layout as well.
This PR also adds some configs and benchmark parameters.

### CODE REVIEW

File: lib/Dialect/TritonGPU/Transforms/OptimizeEpilogue.cpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2023 NVIDIA Corporation & Affiliates. All rights reserved.
```
Comment (zhanglx13) [inline lib/Dialect/TritonGPU/Transforms/OptimizeEpilogue.cpp:2]: @jayfurmanek I remembered you mentioned something about the copyright? Do we use the same as NV?

Comment (alefimov-amd) [reply lib/Dialect/TritonGPU/Transforms/OptimizeEpilogue.cpp:2]: This file is added in IFU, so I think we do not do anything with it.
Or we should rethink our approach and add AMD copyrights into other files we changed.

P.s. By the way, I just realized that there are no copyright in MFMA specific files...

File: python/triton/compiler/compiler.py [hunk 2]
```diff
@@ -78,7 +78,7 @@ def ttir_to_ttgir(mod, num_warps, warpsize):
     return mod
 
 
-def optimize_ttgir(mod, num_stages, arch):
+def optimize_ttgir(mod, num_stages, arch, optimize_epilogue):
```
Comment (alefimov-amd) [inline python/triton/compiler/compiler.py:81]: Is it for easy comparison or we need this flag for tuning?

File: lib/Dialect/TritonGPU/Transforms/OptimizeEpilogue.cpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+      return mlir::failure();
+
+    auto cvtOp = dyn_cast<triton::gpu::ConvertLayoutOp>(val.getDefiningOp());
+    if (!cvtOp)
+      return mlir::failure();
+
+    auto encoding =
+        cvtOp.getSrc().getType().cast<RankedTensorType>().getEncoding();
+    if (!encoding.isa<triton::gpu::MmaEncodingAttr>() &&
+        !encoding.isa<triton::gpu::MfmaEncodingAttr>())
```
Comment (alefimov-amd) [inline lib/Dialect/TritonGPU/Transforms/OptimizeEpilogue.cpp:76]: Can you wrap this line into `#ifdef USE_ROCM`?

**Comment by zhanglx13** [comment]
Performance improvements from this PR:

bs | nheads | seqlen | triton-mlir | bypassLDS | speedup
-- | -- | -- | -- | -- | --
4 | 48 | 1024 | 93 | 97 | 1.04
4 | 48 | 2048 | 103 | 108 | 1.05
4 | 48 | 4096 | 109 | 114 | 1.05
4 | 48 | 8192 | 112 | 117 | 1.04
4 | 48 | 16384 | 113 | 118 | 1.04

**Review by zhanglx13 — APPROVED** [review]

**Pull request merged**